### ISSUE TITLE
determineActualTypes produce too many variants for nested objects

### ISSUE THREAD

**Issue by nkrkv** [sanctuary-js/sanctuary-def#119]
I’ve got a problem when a `TypeError` tries to be too smart. Consider an example:

```javascript

const R = require('ramda');
const $ = require('sanctuary-def');

//-----------------------------------------------------------------------------
//
// Type utilities
//
//-----------------------------------------------------------------------------

const NullaryType = (typeName, predicate) => $.NullaryType(
  `example/${typeName}`,
  `http://example.com/#${typeName}`,
  predicate
);

// :: Type -> (Any -> Boolean)
const hasType = $.test([]);

// :: [Type] -> (Any -> Boolean)
const hasOneOfType = types => R.anyPass(
  R.map(hasType, types)
);

const Model = (typeName, schema) => NullaryType(
  typeName,
  hasType($.RecordType(schema))
);

const OneOfType = (typeName, types) => NullaryType(
  typeName,
  hasOneOfType(types)
);

const AliasType = (typeName, type) => NullaryType(
  typeName,
  hasType(type)
);

//-----------------------------------------------------------------------------
//
// Domain types
//
//-----------------------------------------------------------------------------

const ObjectWithId = NullaryType('ObjectWithId', R.has('id'));

const Identifier = AliasType('Identifier', $.String);

const NodePosition = Model('NodePosition', {
  x: $.Number,
  y: $.Number,
});

const PinRef = Model('PinRef', {
  nodeId: Identifier,
  pinKey: Identifier,
});

const Node = Model('Node', {
  id: Identifier,
  position: NodePosition,
});

const Link = Model('Link', {
  id: Identifier,
  input: PinRef,
  output: PinRef,
});

const Patch = Model('Patch', {
  nodes: $.StrMap(Node),
  links: $.StrMap(Link),
});

const NodeOrId = OneOfType('NodeOrId', [Identifier, ObjectWithId]);
const LinkOrId = OneOfType('LinkOrId', [Identifier, ObjectWithId]);

//-----------------------------------------------------------------------------
//
// Environment
//
//-----------------------------------------------------------------------------

const env = $.env.concat([
  Identifier,
  Link,
  LinkOrId,
  Node,
  NodeOrId,
  Patch,
  PinRef,
]);

const def = $.create({ checkTypes: true, env });

//-----------------------------------------------------------------------------
//
// Main
//
//-----------------------------------------------------------------------------
const foo = def('foo', {}, [Patch, Patch], x => x);

const patch = ({
  nodes: {
    a: { id: 'a' },
    b: { id: 'b' },
    c: { id: 'c' },
  },
  links: {
    x: {
      id: 'x',
      input: { nodeId: 'b', pinKey: 'x' },
      output: { nodeId: 'a', pinKey: 'x' },
    },
    y: {
      id: 'y',
      input: { nodeId: 'c', pinKey: 'x' },
      output: { nodeId: 'b', pinKey: 'x' },
    },
  },
  impls: {
    js: '// ok',
  },
});

foo(patch);
```

Now, if I’d `node index.js`, I get a `TypeError`:

```
...
TypeError: Invalid value

foo :: Patch -> Patch
       ^^^^^
         1

1)  {"impls": {"js": "// ok"}, "links": {"x": {"id": "x", "input": {"nodeId": "b", "pinKey": "x"}, "output": {"nodeId": "a", "pinKey": "x"}}, "y": {"id": "y", "input": {"nodeId": "c", "pinKey": "x"}, "output": {"nodeId": "b", "pinKey": "x"}}}, "nodes": {"a": {"id": "a"}, "b": {"id": "b"}, "c": {"id": "c"}}} :: Object, StrMap Object, StrMap (StrMap Object), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap LinkOrId), StrMap (StrMap NodeOrId), StrMap (StrMap Object), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap LinkOrId), StrMap (StrMap NodeOrId), StrMap (StrMap Object), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap LinkOrId), StrMap (StrMap NodeOrId), StrMap (StrMap Object), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap LinkOrId), StrMap (StrMap NodeOrId), StrMap (StrMap Object), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap LinkOrId), StrMap (StrMap NodeOrId), StrMap (StrMap Object), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap (StrMap String)), StrMap (StrMap (StrMap Identifier)), StrMap (StrMap (StrMap LinkOrId)), StrMap (StrMap (StrMap NodeOrId)), StrMap (StrMap LinkOrId), StrMap (StrMap NodeOrId)

The value at position 1 is not a member of ‘Patch’.

See http://example.com/#Patch for information about the example/Patch type.
...
```

That’s fine because indeed the `patch` object has missing fields in it. However, `determineActualTypes` produces an x-product of all possible variants and it causes performance problems (on throw) and produces much not-so-useful output. In a slightly more complex case node hangs and crashes with stack overflow.

The problem is around `NodeOrId` and `LinkOrId`. If I’d simply remove them from `env`, the output becomes concise:

```
...
TypeError: Invalid value

foo :: Patch -> Patch
       ^^^^^
         1

1)  {"impls": {"js": "// ok"}, "links": {"x": {"id": "x", "input": {"nodeId": "b", "pinKey": "x"}, "output": {"nodeId": "a", "pinKey": "x"}}, "y": {"id": "y", "input": {"nodeId": "c", "pinKey": "x"}, "output": {"nodeId": "b", "pinKey": "x"}}}, "nodes": {"a": {"id": "a"}, "b": {"id": "b"}, "c": {"id": "c"}}} :: Object, StrMap Object

The value at position 1 is not a member of ‘Patch’.

See http://example.com/#Patch for information about the example/Patch type.
...
```

Am I missing the idea and do something wrong when I want to express the following:

- Some functions accept either plain identifier (a string restricted by a regexp) or an object with `id` field containing that identifier;
- There are many record types having `id` field?

**Comment by davidchambers** [comment]
One thing which has long been a source of confusion is the purpose of the environment. It exists to support polymorphic functions such as `id :: a -> a`. By default, `id({foo: 42, bar: true})` is a type error because `{foo: 42, bar: true}` is not a member of any type in the default environment. If one wishes to accept values of type `{ foo :: Number, bar :: Boolean }` in place of a type variable, one must define the record type and include it in the environment.

For `foo :: Patch -> Patch`, though, there's actually no need for `Patch` to be included in the environment. If one is not defining polymorphic functions the environment can simply be `[]`.

Setting aside polymorphic functions, the environment is still useful for generating error messages in "straightforward" cases. For example:

```javascript
S.match(/def/g);
// ! TypeError: Invalid value
//
//   match :: NonGlobalRegExp -> String -> Maybe { groups :: Array (Maybe String), match :: String }
//            ^^^^^^^^^^^^^^^
//                   1
//
//   1)  /def/g :: RegExp, GlobalRegExp
//
//   The value at position 1 is not a member of ‘NonGlobalRegExp’.
//
//   See https://github.com/sanctuary-js/sanctuary-def/tree/v0.9.0#NonGlobalRegExp for information about the sanctuary-def/NonGlobalRegExp type.
```

The fact that `GlobalRegExp` is listed as one of the types of `/def/g` is very much a good thing. So we're in a tricky position, because sometimes enumerating the types is helpful but in other times it generates a huge amount of noise. Perhaps we should truncate the list of types:

    1)  42 :: Number, FiniteNumber, NonZeroFiniteNumber, ... (7 types omitted)

Do you have any suggestions, @nailxx?

**Comment by Avaq** [comment]
> In a slightly more complex case node hangs and crashes with stack overflow. -- @nailxx 

> Perhaps we should truncate the list of types -- @davidchambers 

I think the stack overflow is a more pressing issue to solve. If truncating the list of types to `n` would allow the stack to not exceed some function of `n`, I wholeheartedly support that idea.

**Comment by rjmk** [comment]
I agree both for stack issues and for usability. I wonder whether `create` would need to take a `verbose` option then.

Will truncating the list be sufficient to avoid blowing the stack?

**Comment by nailxx** [comment]
> For foo :: Patch -> Patch, though, there's actually no need for Patch to be included in the environment. If one is not defining polymorphic functions the environment can simply be [].

Thanks for the clarification. Although it is not the case for my simplified example, I actually want my records to appear in polymorphic functions not mentioned here. So this quick fix would not solve the issue.

> Perhaps we should truncate the list of types

That’s a very good idea. And as @Avaq and @rjmk say, the stack overflow is actually a more important problem.

**Comment by Avaq** [comment]
I know I saw the issue before! I think this issue was rediscovered in #158 and solved in #159

**Comment by davidchambers** [comment]
Right you are, Aldwin!
